yarn-project/simulator/src/avm/journal/nullifiers.ts

```diff
@@ -0,0 +1,158 @@
+
+import type { NullifiersDB } from '../../index.js';
+
+/**
+ * A class to manage new nullifier staging and existence checks during a contract call's AVM simulation.
+ * Maintains a pending nullifier cache, and ensures that existence checks fall back to the correct source.
+ * When a contract call completes, its pending nullifier set can be merged into its parent's.
+ */
+export class Nullifiers {
+  /** Cached nullifiers. */
+  private cache: PendingNullifiers;
+  /** Parent's nullifier cache. Checked on cache-miss. */
+  private readonly parentCache: PendingNullifiers | undefined;
+  /** Reference to node storage. Checked on parent cache-miss. */
+  private readonly hostNullifiers: NullifiersDB;
+
+  constructor(hostNullifiers: NullifiersDB, parent?: Nullifiers) {
+    this.hostNullifiers = hostNullifiers;
+    this.parentCache = parent?.cache;
+    this.cache = new PendingNullifiers();
+  }
+
+  /**
+   * Get a nullifier's leaf index and existence status.
+   * 1. Check cache.
+   * 2. Check parent's cache.
+   * 3. Fall back to the host state.
+   * 4. Not found! Nullifier does not exist.
+   *
+   * @param storageAddress - the address of the contract whose storage is being read from
+   * @param nullifier - the nullifier to check for
+   * @returns exists: whether the nullifier exists at all,
+   *          isPending: whether the nullifier was found in a cache,
+   *          leafIndex: the nullifier's leaf index if it exists and is not pending.
+   */
+  public async getNullifierIndex(
+    storageAddress: Fr,
+    nullifier: Fr,
+  ): Promise<[/*exists=*/ boolean, /*isPending=*/ boolean, /*leafIndex=*/ Fr]> {
+    // First check this pending nullifiers
+    let existsAsPending = this.cache.exists(storageAddress, nullifier);
+    // Then check parent's pending nullifiers
+    if (!existsAsPending && this.parentCache) {
+      existsAsPending = this.parentCache?.exists(storageAddress, nullifier);
+    }
+    // Finally try the host's Aztec state (a trip to the database)
+    // If the value is found in the database, it will be associated with a leaf index!
+    let leafIndex: bigint | undefined = undefined;
+    if (!existsAsPending) {
+      leafIndex = await this.hostNullifiers.getNullifierIndex(nullifier);
+    }
+    const exists = existsAsPending || leafIndex !== undefined;
+    leafIndex = leafIndex === undefined ? BigInt(0) : leafIndex;
+    return Promise.resolve([exists, existsAsPending, new Fr(leafIndex)]);
+  }
+
+  /**
+   * Stage a new nullifier (append it to the cache).
+   *
+   * @param storageAddress - the address of the contract that the nullifier is associated with
+   * @param nullifier - the nullifier to stage
+   */
+  public async append(storageAddress: Fr, nullifier: Fr) {
+    const [exists, ,] = await this.getNullifierIndex(storageAddress, nullifier);
+    if (exists) {
+      throw new Error(`Nullifier ${nullifier} at contract ${storageAddress} already exists in parent cache or host.`);
```

Comment:
Is the consequence of this error that the call will revert? If so this should be an AvmExecutionError ?

Reply:
Yeah, I need to give some thought to how we will handle a revert from here. Feels odd for an `AvmExecutionError` to get thrown all the way down here. Maybe the `EMITNULLIFIER` instruction should check existence before it appends, and error if the nullifier exists?

Reply:
I added an option to this function `skipCollisionCheck`. When `append` is actually called as part of the `EMITNULLIFIER` opcode's execution, it now explicitly checks for the nullifier's existence _before calling `append`_ and reverts if found.

I think this is better than throwing an error deep down inside of this module and expecting the AVM to handle it gracefully.

Reply:
it may still be worth having the invariant
as discussed the error can be mapped on the way up